File: java/system_design/CircuitBreaker.java
import java.util.*;
/**
 * CircuitBreaker - [Brief description]
 *
 * <p>OVERVIEW:
 * [Detailed explanation of what this class does]
 *
 * <p>COMPLEXITY:
 * <ul>
 *   <li>Time: [See method documentation]</li>
 *   <li>Space: O(n) where n is [the element count]</li>
 * </ul>
 *
 * <p>USAGE:
 * [How to use this class, with example]
 *
 * @author Interview Preparation
 * @since 1.0
 */

/**
 * [Brief description]
 *
 * @param [param] [description]
 * @return [description]
 * @time O([complexity])
 */
public class CircuitBreaker{enum State{CLOSED,OPEN,HALF_OPEN}State state=State.CLOSED;int failures=0;void fail(){failures++;if(failures>=5)state=State.OPEN;}public static void main(String[]a){CircuitBreaker cb=new CircuitBreaker();for(int i=0;i<5;i++)cb.fail();}}

File: python/system_design/circuit_breaker.py
from enum import Enum
class State(Enum): CLOSED=1; OPEN=2; HALF_OPEN=3
class CircuitBreaker:
    def __init__(self): self.state=State.CLOSED; self.failures=0; self.threshold=5
    def call(self): return self.state==State.CLOSED
    def fail(self): self.failures+=1; self.state=State.OPEN if self.failures>=self.threshold else self.state
if __name__ == "__main__": cb=CircuitBreaker(); [cb.fail() for _ in range(5)]; print(cb.state)